### ISSUE TITLE
grandchildren aggregation broken in ES 2.0?

### ISSUE THREAD

**Issue by chintan-shah-25** [elastic/elasticsearch#15413]
I am looking into this [SO question](http://stackoverflow.com/questions/34222753/elasticsearch-deep-children-aggregation-with-sum-metric-returning-empty-re), I am also facing something similar in my own project after I upgraded from 1.7 to 2.0. **children aggregation** gives weird results.

I created test index like this

```
POST companies
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 0
  },
  "mappings": {
    "category": {
      "properties": {
        "id": {
          "type": "string"
        },
        "title": {
          "type": "string"
        }
      }
    },
    "subcategory": {
      "_parent": {
        "type": "category"
      },
      "properties": {
        "id": {
          "type": "string"
        },
        "title": {
          "type": "string"
        }
      }
    },
    "item": {
      "_parent": {
        "type": "subcategory"
      },
      "properties": {
        "id": {
          "type": "string"
        },
        "price": {
          "type": "integer"
        }
      }
    }
  }
}
```

I indexed some documents
```
PUT companies/category/cat1
{
  "id" : "cat1",
  "title" : "t1"
}

PUT companies/category/cat2
{
  "id" : "cat2",
  "title" : "t2"
}

PUT companies/category/cat3
{
  "id" : "cat3",
  "title" : "t3"
}

PUT companies/subcategory/sub1?parent=cat1
{
  "id" : "sub1",
  "title" : "st1"
}

PUT companies/subcategory/sub2?parent=cat1
{
  "id" : "sub2",
  "title" : "st2"
}

PUT companies/subcategory/sub3?parent=cat2
{
  "id" : "sub3",
  "title" : "st2"
}

PUT companies/subcategory/sub4?parent=cat2
{
  "id" : "sub4",
  "title" : "st4"
}

PUT companies/subcategory/sub5?parent=cat3
{
  "id" : "sub5",
  "title" : "st4"
}

PUT companies/subcategory/sub6?parent=cat3
{
  "id" : "sub6",
  "title" : "st6"
}

PUT companies/subcategory/sub7?parent=cat3
{
  "id" : "sub7",
  "title" : "st7"
}

PUT companies/item/i1?parent=sub1&routing=cat1
{
  "id" : "i1",
  "price" : 100
}

PUT companies/item/i2?parent=sub1&routing=cat1
{
  "id" : "i2",
  "price" : 200
}

PUT companies/item/i3?parent=sub2&routing=cat1
{
  "id" : "i3",
  "price" : 200
}

PUT companies/item/i4?parent=sub2&routing=cat1
{
  "id" : "i4",
  "price" : 150
}

PUT companies/item/i5?parent=sub3&routing=cat2
{
  "id" : "i5",
  "price" : 150
}

PUT companies/item/i6?parent=sub3&routing=cat2
{
  "id" : "i6",
  "price" : 180
}

PUT companies/item/i7?parent=sub4&routing=cat2
{
  "id" : "i7",
  "price" : 180
}

PUT companies/item/i8?parent=sub5&routing=cat3
{
  "id" : "i8",
  "price" : 1180
}

PUT companies/item/i9?parent=sub6&routing=cat3
{
  "id" : "i9",
  "price" : 11180
}

PUT companies/item/i10?parent=sub7&routing=cat3
{
  "id" : "i10",
  "price" : 180
}

PUT companies/item/i11?parent=sub7&routing=cat3
{
  "id" : "i11",
  "price" : 150
}

PUT companies/item/i12?parent=sub7&routing=cat3
{
  "id" : "i12",
  "price" : 210
}

```

This is the aggregation query
```
GET companies/_search
{
  "query": {
    "has_child": {
      "type": "subcategory",
      "query": {
        "has_child": {
          "type": "item",
          "query": {
            "range": {
              "price": {
                "gte": 100,
                "lte": 250
              }
            }
          }
        }
      }
    }
  },
  "aggs": {
    "categories": {
      "terms": {
        "field": "id"
      },
      "aggs": {
        "sub_categories": {
          "children": {
            "type": "subcategory"
          },
          "aggs": {
            "sub_category_ids": {
              "terms": {
                "field": "id"
              },
              "aggs": {
                "items": {
                  "children": {
                    "type": "item"
                  },
                  "aggs": {
                    "price": {
                      "sum": {
                        "field": "price"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  "size": 0
}
```

Results for ES 1.7
```
{
   "took": 189,
   "timed_out": false,
   "_shards": {
      "total": 1,
      "successful": 1,
      "failed": 0
   },
   "hits": {
      "total": 3,
      "max_score": 0,
      "hits": []
   },
   "aggregations": {
      "categories": {
         "doc_count_error_upper_bound": 0,
         "sum_other_doc_count": 0,
         "buckets": [
            {
               "key": "cat1",
               "doc_count": 1,
               "sub_categories": {
                  "doc_count": 2,
                  "sub_category_ids": {
                     "doc_count_error_upper_bound": 0,
                     "sum_other_doc_count": 0,
                     "buckets": [
                        {
                           "key": "sub1",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 2,
                              "price": {
                                 "value": 300
                              }
                           }
                        },
                        {
                           "key": "sub2",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 2,
                              "price": {
                                 "value": 350
                              }
                           }
                        }
                     ]
                  }
               }
            },
            {
               "key": "cat2",
               "doc_count": 1,
               "sub_categories": {
                  "doc_count": 2,
                  "sub_category_ids": {
                     "doc_count_error_upper_bound": 0,
                     "sum_other_doc_count": 0,
                     "buckets": [
                        {
                           "key": "sub3",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 2,
                              "price": {
                                 "value": 330
                              }
                           }
                        },
                        {
                           "key": "sub4",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 1,
                              "price": {
                                 "value": 180
                              }
                           }
                        }
                     ]
                  }
               }
            },
            {
               "key": "cat3",
               "doc_count": 1,
               "sub_categories": {
                  "doc_count": 3,
                  "sub_category_ids": {
                     "doc_count_error_upper_bound": 0,
                     "sum_other_doc_count": 0,
                     "buckets": [
                        {
                           "key": "sub5",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 1,
                              "price": {
                                 "value": 1180
                              }
                           }
                        },
                        {
                           "key": "sub6",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 1,
                              "price": {
                                 "value": 11180
                              }
                           }
                        },
                        {
                           "key": "sub7",
                           "doc_count": 1,
                           "items": {
                              "doc_count": 3,
                              "price": {
                                 "value": 540
                              }
                           }
                        }
                     ]
                  }
               }
            }
         ]
      }
   }
}
```

Result for ES 2.0

```
{
  "took": 24,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "hits": {
    "total": 3,
    "max_score": 1,
    "hits": [
      {
        "_index": "companies",
        "_type": "category",
        "_id": "cat1",
        "_score": 1,
        "_source": {
          "id": "cat1",
          "title": "t1"
        }
      },
      {
        "_index": "companies",
        "_type": "category",
        "_id": "cat2",
        "_score": 1,
        "_source": {
          "id": "cat2",
          "title": "t2"
        }
      },
      {
        "_index": "companies",
        "_type": "category",
        "_id": "cat3",
        "_score": 1,
        "_source": {
          "id": "cat3",
          "title": "t3"
        }
      }
    ]
  },
  "aggregations": {
    "categories": {
      "doc_count_error_upper_bound": 0,
      "sum_other_doc_count": 0,
      "buckets": [
        {
          "key": "cat1",
          "doc_count": 1,
          "sub_categories": {
            "doc_count": 2,
            "sub_category_ids": {
              "doc_count_error_upper_bound": 0,
              "sum_other_doc_count": 0,
              "buckets": [
                {
                  "key": "sub1",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                },
                {
                  "key": "sub2",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                }
              ]
            }
          }
        },
        {
          "key": "cat2",
          "doc_count": 1,
          "sub_categories": {
            "doc_count": 2,
            "sub_category_ids": {
              "doc_count_error_upper_bound": 0,
              "sum_other_doc_count": 0,
              "buckets": [
                {
                  "key": "sub3",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                },
                {
                  "key": "sub4",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                }
              ]
            }
          }
        },
        {
          "key": "cat3",
          "doc_count": 1,
          "sub_categories": {
            "doc_count": 3,
            "sub_category_ids": {
              "doc_count_error_upper_bound": 0,
              "sum_other_doc_count": 0,
              "buckets": [
                {
                  "key": "sub5",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 1,
                    "price": {
                      "value": 1180
                    }
                  }
                },
                {
                  "key": "sub6",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                },
                {
                  "key": "sub7",
                  "doc_count": 1,
                  "items": {
                    "doc_count": 0,
                    "price": {
                      "value": 0
                    }
                  }
                }
              ]
            }
          }
        }
      ]
    }
  }
}
```

Did I miss something?

**Comment by clintongormley** [comment]
Hmm I've tried this on 2.0.0 and 2.1.0, and with different numbers of shards.  I'm unable to replicate this.  Could you confirm that this recreation is correct?

**Comment by clintongormley** [comment]
Related to https://github.com/elastic/elasticsearch/issues/15005#issuecomment-163090273 ?

**Comment by chintan-shah-25** [comment]
I recreated index after deleting  and It is not working with ES 2.0, This is another [SO question](http://stackoverflow.com/questions/34072977/children-aggregation-broken-in-elasticsearch-2-1) and I also tried that on both ES 1.7 and 2.0 and it is not working on 2.0.

**Comment by clintongormley** [comment]
@chintan-shah-25 by "not working" do you mean that the recreation isn't working? In other words, Elasticsearch is working correctly?

**Comment by chintan-shah-25** [comment]
I meant query is not working, it is still giving wrong results in 2.0, also could you please try the other question in the link, it is also giving wrong results in 2.0 but working perfectly in 1.7

**Comment by clintongormley** [comment]
@chintan-shah-25 For some reason, your recreation works fine for me, but the one you linked to on SO shows the error on 2.0 and 2.1:

    PUT /test_index
    {
      "mappings": {
        "parent_doc": {
          "properties": {
            "name": {
              "type": "string",
              "index": "not_analyzed"
            },
            "town": {
              "type": "string",
              "index": "not_analyzed"
            }
          }
        },
        "child_doc": {
          "_parent": {
            "type": "parent_doc"
          },
          "properties": {
            "name": {
              "type": "string",
              "index": "not_analyzed"
            },
            "age": {
              "type": "integer"
            }
          }
        }
      }
    }
    
    POST /test_index/parent_doc/_bulk
    {"index":{"_id":1}}
    {"name":"Bob","town":"Memphis"}
    {"index":{"_id":2}}
    {"name":"Alice","town":"Chicago"}
    {"index":{"_id":3}}
    {"name":"Bill","town":"Chicago"}
    
    POST /test_index/child_doc/_bulk
    {"index":{"_id":3,"_parent":1}}
    {"name":"Jill","age":5}
    {"index":{"_id":4,"_parent":1}}
    {"name":"Joey","age":3}
    {"index":{"_id":5,"_parent":2}}
    {"name":"John","age":2}
    {"index":{"_id":6,"_parent":3}}
    {"name":"Betty","age":6}
    {"index":{"_id":7,"_parent":3}}
    {"name":"Dan","age":1}
    
    
    POST /test_index/parent_doc/_search
    {
      "size": 0,
      "aggs": {
        "towns": {
          "terms": {
            "field": "town"
          },
          "aggs": {
            "parent_names": {
              "terms": {
                "field": "name"
              },
              "aggs": {
                "child_docs": {
                  "children": {
                    "type": "child_doc"
                  }
                }
              }
            }
          }
        }
      }
    }

**Comment by chintan-shah-25** [comment]
Is it throwing error or giving wrong results?

**Comment by clintongormley** [comment]
giving wrong results

**Comment by martijnvg** [comment]
@chintan-shah-25 @clintongormley The children agg makes an assumption (that all segments are being seen by children agg) that was true in 1.x but not in 2.x. I opened #15457 which fixes that.

**Comment by martijnvg** [comment]
This bug has been fixed via #15457 and back ported to the 2.x, 2.1 and 2.0 branches.

**Comment by chintan-shah-25** [comment]
Thanks a lot @martijnvg @clintongormley :)
